### PR TITLE
sleep-config: cleanup round three

### PR DESCRIPTION
**Pull request by YHNdnzj**
This should be the last round that needs to complete before v255-rc1 :)

### CODE REVIEW

File: src/shared/sleep-config.c [hunk 1]
```diff
@@ -65,12 +102,11 @@ int parse_sleep_config(SleepConfig **ret) {
                 { "Sleep", "AllowSuspendThenHibernate", config_parse_tristate, 0, &allow_s2h                      },
                 { "Sleep", "AllowHybridSleep",          config_parse_tristate, 0, &allow_hybrid_sleep             },
 
-                { "Sleep", "SuspendMode",               config_parse_strv,     0, sc->modes + SLEEP_SUSPEND       },
                 { "Sleep", "SuspendState",              config_parse_strv,     0, sc->states + SLEEP_SUSPEND      },
+
                 { "Sleep", "HibernateMode",             config_parse_strv,     0, sc->modes + SLEEP_HIBERNATE     },
-                { "Sleep", "HibernateState",            config_parse_strv,     0, sc->states + SLEEP_HIBERNATE    },
```
Comment (bluca) [inline src/shared/sleep-config.c:71]: As far as I can tell these have been around for years, so can't just remove them - instead print a deprecated warning and ignore them if they are set

**Review by bluca — APPROVED** [review]

File: src/hibernate-resume/hibernate-resume.c [hunk 2]
```diff
@@ -12,7 +12,7 @@
 #include "parse-util.h"
 #include "static-destruct.h"
 
-HibernateInfo arg_info = {};
+static HibernateInfo arg_info = {};
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume.c:15]: weird that the compiler didn't complain about this

File: man/systemd-sleep.conf.xml [hunk 3]
```diff
@@ -162,6 +160,9 @@
         url="https://www.kernel.org/doc/html/latest/admin-guide/pm/sleep-states.html#basic-sysfs-interfaces-for-system-suspend-and-hibernation">the
         kernel documentation</ulink> for more details.</para>
 
+        <para>Note that hybrid sleep corresponds to the <literal>suspend</literal> disk mode. If <varname>HybridSleepMode=</varname>
+        is overridden, you might get plain hibernation instead.</para>
```
Comment (poettering) [inline man/systemd-sleep.conf.xml:164]: that smells as if we shouldn't allow setting HybridSleepMode= either

File: src/shared/sleep-config.c [hunk 4]
```diff
@@ -33,18 +33,55 @@ static const char* const sleep_operation_table[_SLEEP_OPERATION_MAX] = {
...
 
+static void sleep_config_validate_state_and_mode(SleepConfig *sc) {
+        assert(sc);
+
+        /* So we should really not allow setting SuspendState= to 'disk', which means hibernation. We have
+         * SLEEP_HIBERNATE for proper hibernation support, which includes checks for resume support (through
+         * EFI variable or resume= kernel command line option). It's simply not sensible to call the suspend
+         * operation but eventually do an unsafe hibernation. */
+        if (strv_contains(sc->states[SLEEP_SUSPEND], "disk")) {
+                strv_remove(sc->states[SLEEP_SUSPEND], "disk");
```
Comment (poettering) [inline src/shared/sleep-config.c:68]: frankly, this all looks as if this shouldn't be an strv in the first place. But a bit mask, i.e.:

```c
typedef enum SleepState {
        SLEEP_MEM, 
        SLEEP_STANDBY,
        SLEEP_FREEZE,
        SLEEP_DISK,
        _SLEEP_STATE_MAX,
        _SLEEP_STATE_INVALID = -EINVAL,
}; 
```

and similar for mode.

and then we should have tables somewhere:

```c
static const unsigned sleep_state_defaults[_SLEEP_OPERATION_MAX] = {
        [SLEEP_SUSPEND] = (1U << SLEEP_MEM) | (1U << SLEEP_STANDBY) | (1U << SLEEP_FREEZE),
        …
};

static const unsigned sleep_state_allowed[_SLEEP_OPERATION_MAX] = {
        [SLEEP_SUSPEND] = …
}
```

And then we default to the entry in the upper table, and only allow people to set the masks in the lower table. You get the idea.

**Comment by poettering** [comment]
lgtm. i guess we can merge this. the further clean-ups i suggested can be done later (if anyone is interested).

**Comment by poettering** [comment]
btw, another clean-up i'd like to see one day: I think it was a mistake to let the users/desktops make the choice between suspend/hibernate/s2h. I think we should try to fix that though. Specifically: add a new Sleep() bus call (and `systemctl sleep` command) which will redirect to Suspend()/Hibernate()/SuspendThenHibernate() depending on what is available (or configured). Should probably by default use SuspendThenHibernate() if it is supported, and Suspend() otherwise, and finally Hibernate() as last fallback. And people should be to configure this explicitly in a new SleepOperation= setting in logind.conf or so.

**Comment by YHNdnzj** [comment]
> lgtm. i guess we can merge this as it is. the further clean-ups i suggested can be done later (if anyone is interested).

I have at least two more cleanups queued up for `sleep`. Will add this to my to-do list :) Just want to get this compat break in before rc1.

File: src/shared/sleep-config.c (hunk 4 above)
Comment (YHNdnzj) [reply src/shared/sleep-config.c:68]: Hmm, actually we can't do that. Currently, we try the given states/modes in order. But if we switch to something like this it's no longer possible.

File: man/systemd-sleep.conf.xml (hunk 3 above)
Comment (YHNdnzj) [reply man/systemd-sleep.conf.xml:164]: -> #29688

**Pull request merged**